src/sentry/services/hybrid_cloud/organizationmapping.py

```diff
@@ -0,0 +1,141 @@
+from abc import abstractmethod
+from dataclasses import dataclass, fields
+
+from django.db import IntegrityError, transaction
+from rest_framework import status
+
+from sentry.models.organizationmapping import OrganizationMapping
+from sentry.models.user import User
+from sentry.services.hybrid_cloud import InterfaceWithLifecycle, silo_mode_delegation
+from sentry.silo import SiloMode
+from sentry.silo.client import ControlSiloClient
+
+
+@dataclass(frozen=True, eq=True)
+class APIOrganizationMapping:
+    id: int = -1
+    organization_id: int = -1
+    slug: str = ""
+    region_name: str = ""
+
+
+class OrganizationMappingService(InterfaceWithLifecycle):
+    @abstractmethod
+    def create(
+        self,
+        user: User,
+        organization_id: int,
+        slug: str,
+        stripe_id: str,
+        idempotency_key: str,
+        region_name: str,
+    ) -> APIOrganizationMapping:
+        """
+        This method returns a new or recreated OrganizationMapping object.
+        Will raise IntegrityError if the slug already exists.
+        :param organization_id:
+        The org id to create the slug for
+        :param slug:
+        A slug to reserve for this organization
+        :param stripe_id:
+        A unique per customer stripe identifier
+        :return:
+        """
+        pass
+
+    # TODO: Implement update that allows setting stripe ID for new orgs
+
+
+class DatabaseBackedOrganizationMappingService(OrganizationMappingService):
+    def create(
+        self,
+        user: User,
+        organization_id: int,
+        slug: str,
+        stripe_id: str,
+        idempotency_key: str,
+        region_name: str,
+    ) -> APIOrganizationMapping:
+        with transaction.atomic():
+            try:
+                # Creating an identical mapping should succeed, even if a record already exists
+                # with this slug. We allow this IFF the idempotency key is identical
+                mapping = OrganizationMapping.objects.create(
+                    organization_id=organization_id,
+                    slug=slug,
+                    stripe_id=stripe_id,
+                    idempotency_key=idempotency_key,
+                    region_name=region_name,
+                )
+                return self.serialize_organization_mapping(mapping)
+            except IntegrityError:
+                pass
+
+        # If we got here, the slug already exists
+        if idempotency_key != "":
+            try:
+                existing = OrganizationMapping.objects.get(
+                    slug=slug, idempotency_key=idempotency_key
+                )
+                existing.organization_id = organization_id
```

Comment:
Hmmm, I don't think the idempotency key works well.  I think you don't even need it -- all you really need is to query `OrganizationMapping.objects.get(slug=slug, organization_id=organization_id)` right?  Like, idempotency is about the same exact org reserving the same exact slug -- that is totes safe.

As written, the idempotency_key has an implicit requirement that it be stable (otherwise future requests do fail, even if it is the same org updating to the same slug), but that feels odd.  I just think it just isn't needed in the end.

Reply:
Ah you're thinking about the update case. That might be true, but in the _create_ case we need the extra info.

Reply:
Ok, yeah, that makes total sense.  On the update case, if we can rely on the organization id, that would be ideal, but if the idempotency_key is stable it also works fine.